### ISSUE TITLE
Helm 3: action.Install and action.Upgrade have different mechanisms to set values

### ISSUE THREAD

**Issue by joelanford** [helm/helm#5827]
Using the Helm 3 libraries, I'm looking for a consistent way to populate an action struct's chart values from an existing `map[string]interface{}`.

### Install 

`install.Run` uses `i.rawValues` (a `map[string]interface{}`) when calling `ToRenderValues`

https://github.com/helm/helm/blob/2a3b7b5395bb7d71919ed3592c8c63c425353851/pkg/action/install.go#L128

`action.Install` exposes `Values` as `[]string` from its embedded `ValuesOptions` struct. It's `rawValues` is unexported.

https://github.com/helm/helm/blob/2a3b7b5395bb7d71919ed3592c8c63c425353851/pkg/action/install.go#L65-L89

So to populate values for an install, it looks like I have to:
1. Marshal my `map[string]interface{}` value out to a temporary YAML file
2. Set `install.ValueFiles`
3. Call `install.ValueOptions.MergeValues`
4. Call `install.Run`
5. Cleanup the temporary file.

### Upgrade

`upgrade.Run` uses `u.Values` when calling `ToRenderValues` (and seems to ignore `u.rawValues`?)

https://github.com/helm/helm/blob/2a3b7b5395bb7d71919ed3592c8c63c425353851/pkg/action/upgrade.go#L157

So since `Values` is exported publicly as a `map[string]interface{}` (shadowing `ValuesOptions.Values`),  it looks like I can set `upgrade.Value` directly, and everything seems to work.

https://github.com/helm/helm/blob/2a3b7b5395bb7d71919ed3592c8c63c425353851/pkg/action/upgrade.go#L37-L59

Just want to make sure I'm understanding things correctly and to see if this discrepancy is intended :slightly_smiling_face:

Thanks!

**Comment by rokii** [comment]
This causes #5792

**Comment by joelanford** [comment]
@rokii Thanks for pointing that out. It looks like your PR solves the issue with values not being set properly during upgrade, but it doesn't solve my primary issue, which is to have a way to set the raw values directly.

As far as I can tell, it looks like one way to do that would be to export `ValueOptions.rawValues`. so that I can set it directly without having to marshal a `map[string]interface{}` into a file.

Does that sound like a reasonable solution? If so, would you want to add that to #5824?

**Comment by rokii** [comment]
Exporting `ValueOptions.rawValues` directly should be a new feature. 
I think it needs the owner team's agreement. @adamreese

**Comment by bacongobbler** [comment]
`ValueOptions.rawValues` is intended to be an internal variable that was originally designed to not be exposed to the end-user. We should definitely have a way to inject new values from external sources. I'll have a look at #5894 and see if that already exists (and if not, how we should add that). `.Values` was intended to be the injection site for new values.

**Comment by bacongobbler** [comment]
Thanks @rokii for fixing this discrepancy.

@joelanford moving forward, the process for merging values should be the same as `helm install`:

1. Set `upgrade.ValueFiles`, `upgrade.Values` or `upgrade.StringValues`
2. Call `upgrade.ValueOptions.MergeValues`
3. Call `upgrade.Run`

Thanks for the report!

**Comment by joelanford** [comment]
@bacongobbler with #5824 merged, the install and upgrade mechanisms are now consistent 🎉

However, if I understand correctly, it now also makes it awkward to build the release values when the input format is already a `map[string]interface{}` (e.g. a Kubernetes custom resource `spec` used to define the complete set of user-defined release values) and not a list of files or dot-delimited values. As I mentioned, it looks like my options now are either:

1. Marshal out the map of values to a temp file and set `upgrade.ValueFiles` to `[]string{tempFile}`
2. Write a function that converts `map[string]interface{}` to the dot-delimited strings expected by `upgrade.Values` and `upgrade.StringValues`

Neither of these seem like the right thing to do because `ValueOptions.MergeValues` is turning things right back back into a `map[string]interface{}`.

Would you be opposed to the addition of a `ValueOptions.ValueMaps []map[string]interface{}` field and updating `MergeValues` with something like:

```go
func (v *ValueOptions) MergeValues(settings cli.EnvSettings) error {
	base := map[string]interface{}{}

	for _, valueMap := range v.ValueMaps {
		base = mergeMaps(base, valueMap)
	}

	...

	v.rawValues = base
	return nil
}
```

**Comment by bacongobbler** [comment]
So after having a face-to-face, there definitely seems to be some open design questions about how values are merged from the CLI and how those values should be handled when being passed into the SDK.

From a first glance it seems like `MergeValues` and the three input parameters should be a concern of the CLI, but not a concern for the SDK user. If we could refactor `MergeValues` to return a `map[string]interface{}` that can be passed to the actual SDK calls, that would likely be a better move forward for both parties: the CLI doesn't have to change, but it would enable use cases like @joelanford's to pass in values parsed from a separate API.

**Comment by joelanford** [comment]
Thanks @bacongobbler! That sounds like an even better idea.

**Comment by bacongobbler** [comment]
@joelanford have you had a chance to take a closer look at this and see if the proposed changes in my previous comment seems feasible? I've been busy with other work so I haven't been able to take a closer look at this. I'd love to get your eyes on this if possible, given you're ultimately the end user consuming the API you'd be more well-equipped to determine if the end result works for you. :)

**Comment by joelanford** [comment]
I haven't had a chance either. I've been on vacation for a few weeks and I'm busy this week. But I should have more time to take a look next week. If it looks straightforward, I'll send up a PR.

**Comment by joelanford** [comment]
@bacongobbler It looks like #6008 solves the problem by introducing a constructor that sets `rawValues` directly. That solves the original problem, but not in a way that decouples the SDK and CLI.

I think I prefer the approach you described (where `MergeValues` returns a `map[string]interface{}`), but I think that would mean removing `rawValues` from the `ValuesOptions` struct, which would also mean removing the constructor since there would be nothing else for it to do.

Just wanted your thoughts on this before I work on a PR.

**Comment by bacongobbler** [comment]
as long as functionality remains the same, SGTM! This is the opportunity to make breaking changes to the API. :)

**Comment by bacongobbler** [comment]
fixed in #6142
